### DISCUSSION TITLE
How to fetch data with authorization?

### DISCUSSION THREAD

**Discussion by Neutronas** [nextauthjs/next-auth#1512] [Help] [answered]
**Your question**
How to get session in the API if it is being fetched from another page?

**What are you trying to do**
I am trying to build a page where personal information would be shown. This page is only loaded if session exists. In this page I am trying to use my new API code, which could only be reached if user is logged in. However, it returns that session is null when API is reached through fetch.

**Reproduction**
I have page that tries to fetch API through (session is available in the page itself)

```
export async function getServerSideProps(context) {
  const res = await fetch(`${process.env.BASE_URL}/api/profile`);
  console.log(res);
}
```

this /api/profile page has handler like this:
```
const handler = async (req, res) => {
  const session = await getSession({req}); // **this returns null**
  if (!session) {
    return res.status(401).send('Not Authorized');
  }
...
```

Session is returned as null once it was called from fetch.

**Feedback**
Documentation lacks information how to create private routes / pages and API that can only be reached while authenticated.

* [ ] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [x] Could not find relevant documentation
* [ ] Found the example project helpful
* [ ] Did not find the example project helpful

**Comment by balazsorban44** [comment, marked as answer]
There are many ways of doing this, here is a potential solution https://github.com/nextauthjs/next-auth/issues/1210#issuecomment-782630909


Please study the docs carefully https://next-auth.js.org/tutorials/securing-pages-and-api-routes


You could also have checked the example project https://github.com/nextauthjs/next-auth-example

NOTE. You probably don't want to call an API route from inside getServerSideProps, just use the logic directly.

**Comment by resulyrt93** [comment]
@balazsorban44 your answer is irrelevant with question. actually in question asking how we can call api route (route handler) from client component. I guess we should add cookies in request to authenticate by api. But is there a elegant way for it?

**Comment by AndonMitev** [comment]
Hello i'm stucking to the same issue i do have next-auth with custom provider and access token everything is saved but then in my service:

```
import { getSession } from 'next-auth/react';
import { NextApiRequest } from 'next';

const baseUrl = 'http://localhost:3004/api/auth/password';

enum API_ERRORS {
  UPDATE_PASSWORD = 'Failed to update password'
}

const getAccessToken = async (req: NextApiRequest) => {
  const session = (await getSession({ req })) as any;
  return session?.accessToken;
};

export const updatePasswordApi = async (data: {
  oldPassword: string;
  newPassword: string;
}) => {
  const accessToken = await getAccessToken();

  const response = await fetch(baseUrl, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer {accessToken}`
    }
  });

  if (!response.ok) {
    throw new Error(API_ERRORS.UPDATE_PASSWORD);
  }

  return response.json();
};

```

really not sure how to extract the access token and pass it to the call, any ideas?
